### PR TITLE
Add com.bitwarden.desktop

### PR DESCRIPTION
**Pull request by ghisvail**
This PR provides a flatpak for the [Bitwarden Destkop Application](https://github.com/bitwarden/desktop).

Unfortunately, I am gonna need some help to bring this one to the finish line. The Electron packaging for this application deviates a bit from the usual documentation, and I can't find the appropriate wrapper script to launch it.

There is a large selection of scripts in [package.json](https://github.com/bitwarden/desktop/blob/a9067f56188b4c69e82ccf2443d085109ad5bb79/package.json#L19), one of which (`electron`) is supposed to be called to run the app. However, the latter still calls build commands, so the execution fails in the sandbox.

Any ideas?

### CODE REVIEW

File: com.bitwarden.desktop.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>com.bitwarden.desktop</id>
```
[the remark this answers is not in the document]
Comment (TingPing) [reply com.bitwarden.desktop.appdata.xml]: Maybe `com.bitwarden.BitwardenDesktop` if that part is important.

File: com.bitwarden.desktop.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
+app-id: com.bitwarden.desktop
+base: org.electronjs.Electron2.BaseApp
+base-version: stable
+runtime: org.freedesktop.Platform
+runtime-version: 1.6
```
Comment (TingPing) [inline com.bitwarden.desktop.yml]: This is a string.

File: com.bitwarden.desktop.appdata.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (x80486) [reply com.bitwarden.desktop.appdata.xml]: I think that's better; the `bitwarden/desktop` repository refers to the artifact as `Bitwarden Desktop Application` — but no need for "application" 😉 in the name.

File: com.bitwarden.desktop.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      - /app/lib/node_modules/npm/changelogs
+      - /app/lib/node_modules/npm/doc
+      - /app/lib/node_modules/npm/html
+      - /app/lib/node_modules/npm/man
+      - /app/lib/node_modules/npm/scripts
+    sources:
+      - type: archive
+        url: https://nodejs.org/dist/v8.11.3/node-v8.11.3.tar.xz
+        sha256: 577c751fdca91c46c60ffd8352e5b465881373bfdde212c17c3a3c1bd2616ee0
+  - name: libsecret
```
Comment (eszlari) [inline com.bitwarden.desktop.yml]: insert newlines between modules: https://github.com/flathub/flathub/wiki/YAML-Style-Guide

File: com.bitwarden.desktop.appdata.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ghisvail) [reply com.bitwarden.desktop.appdata.xml]: I had `com.bitwarden.Bitwarden` originally, but then figured out upstream listed an explicit `appId` in their `package.json`. It depends how strict we are supposed to be with the "use upstream's app-id if it already exists" rule.

If we are fine using a different one, I'd rather use `com.bitwarden.Bitwarden` as I planned originally.

File: com.bitwarden.desktop.yml (hunk 2 above)
Comment (ghisvail) [reply com.bitwarden.desktop.yml]: I don't understand your comment, sorry.

File: com.bitwarden.desktop.yml (hunk 3 above)
Comment (ghisvail) [reply com.bitwarden.desktop.yml]: Oh, I was not aware we had an official style guide.

Thanks for pointing it out :-)

File: com.bitwarden.desktop.appdata.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (nedrichards) [reply com.bitwarden.desktop.appdata.xml]: we should use upstream's ID whereever possible. I think it should be possible here.

**Comment by nedrichards** [comment]
bot, build com.bitwarden.desktop

File: com.bitwarden.desktop.appdata.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ghisvail) [reply com.bitwarden.desktop.appdata.xml]: Agreed.

File: com.bitwarden.desktop.yml (hunk 2 above)
Comment (TingPing) [reply com.bitwarden.desktop.yml]: `runtime-version: '1.6'`

**Comment by ghisvail** [comment]
Builds are failing on arm due to the `desktop-idle` dependency:

```
npm ERR! notsup Unsupported platform for desktop-idle@1.1.1: wanted {"os":"windows,linux,darwin,win32","arch":"x64,ia32"} (current: {"os":"linux","arch":"arm64"})
```

Looks like I'll have no choice but restrict the build architectures.

**Comment by ghisvail** [comment]
I have asked upstream about the arm restriction [here](https://github.com/bithavoc/node-desktop-idle/issues/8). Meanwhile, I'll restrict the builds to x86.

File: com.bitwarden.desktop.yml (hunk 2 above)
Comment (ghisvail) [reply com.bitwarden.desktop.yml]: Why does it work without, I wonder?

**Comment by AsciiWolf** [comment]
bot, build com.bitwarden.desktop

**Comment by ghisvail** [comment]
Thanks @bithavoc.

I'll ask the upstream developers of Bitwarden to bump their version of `desktop-idle`.

**Comment by ghisvail** [comment]
After refreshing the flatpak packaging with the current tip of `master`, the build fails with a different error:

```
Running: npm run build --prefix=main --offline --cache=/run/build/bitwarden/npm-cache --nodedir=/app
npm info it worked if it ends with ok
npm info using npm@5.6.0
npm info using node@v8.11.3
npm info lifecycle bitwarden@0.0.0~prebuild: bitwarden@0.0.0
npm info lifecycle bitwarden@0.0.0~build: bitwarden@0.0.0

> bitwarden@0.0.0 build /run/build/bitwarden/main
> concurrently -n Main,Rend -c yellow,cyan "npm run build:main" "npm run build:renderer"

[Rend] npmnpm info it worked if it ends with ok
[Main] npm info using npm@5.6.0
[Main] npm info using node@v8.11.3
[Rend]  info it worked if it ends with ok
[Rend] npm info using npm@5.6.0
[Rend] npm info using node@v8.11.3
[Main] npm info lifecycle bitwarden@0.0.0~prebuild:main: bitwarden@0.0.0
[Main] npm info lifecycle bitwarden@0.0.0~build:main: bitwarden@0.0.0
[Rend] npm info lifecycle bitwarden@0.0.0~prebuild:renderer: bitwarden@0.0.0
[Rend] npm info lifecycle bitwarden@0.0.0~build:renderer: bitwarden@0.0.0
[Main] 
[Main] > bitwarden@0.0.0 build:main /run/build/bitwarden/main
[Main] > webpack --config webpack.main.js
[Main] 
[Rend] 
[Rend] > bitwarden@0.0.0 build:renderer /run/build/bitwarden/main
[Rend] > gulp prebuild:renderer && webpack --config webpack.renderer.js
[Rend] 
[Rend] [13:45:02] Using gulpfile /run/build/bitwarden/main/gulpfile.js
[Rend] [13:45:02] Starting 'clean'...
[Rend] [13:45:02] Finished 'clean' after 23 ms
[Rend] [13:45:02] Starting 'webfonts'...
[Rend] events.js:183
[Rend]       throw er; // Unhandled 'error' event
[Rend]       ^
[Rend] 
[Rend] Error: getaddrinfo ENOTFOUND fonts.googleapis.com fonts.googleapis.com:443
[Rend]     at errnoException (dns.js:50:10)
[Rend]     at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:92:26)
[Rend] npm info lifecycle bitwarden@0.0.0~build:renderer: Failed to exec build:renderer script
[Rend] npm ERR! code ELIFECYCLE
npm ERR! errno 1
[Rend] npm ERR! bitwarden@0.0.0 build:renderer: `gulp prebuild:renderer && webpack --config webpack.renderer.js`
[Rend] npm ERR! Exit status 1
[Rend] npm ERR! 
npm ERR! Failed at the bitwarden@0.0.0 build:renderer script.
[Rend] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[Rend] 
[Rend] npm ERR! A complete log of this run can be found in:
[Rend] npm ERR!     /run/build/bitwarden/npm-cache/_logs/2018-08-27T11_45_02_444Z-debug.log
[Rend] npm run build:renderer exited with code 1
[Main] clean-webpack-plugin: /run/build/bitwarden/main/build/* has been removed.
[Main] Hash: 816f43ec2d51e66e9199
[Main] Version: webpack 3.12.0
[Main] Time: 10723ms
[Main]                        Asset       Size  Chunks             Chunk Names
[Main]     locales/es/messages.json    26.9 kB          [emitted]  
[Main]                      main.js     101 kB       0  [emitted]  main
[Main]          images/fa-globe.png  344 bytes          [emitted]  
[Main]    images/icon-highlight.png  292 bytes          [emitted]  
[Main] images/icon-highlight@2x.png  438 bytes          [emitted]  
[Main]     images/icon-template.png  292 bytes          [emitted]  
[Main]  images/icon-template@2x.png  438 bytes          [emitted]  
[Main]              images/icon.ico    20.2 kB          [emitted]  
[Main]              images/icon.png    12.3 kB          [emitted]  
[Main]           images/loading.svg  291 bytes          [emitted]  
[Main]      images/logo-dark@2x.png    9.32 kB          [emitted]  
[Main]     images/logo-white@2x.png    9.25 kB          [emitted]  
[Main]      images/two-factor/0.png    6.05 kB          [emitted]  
[Main]      images/two-factor/1.png    2.67 kB          [emitted]  
[Main]      images/two-factor/2.png    1.21 kB          [emitted]  
[Main]      images/two-factor/3.png    1.62 kB          [emitted]  
[Main]      images/two-factor/4.png    4.81 kB          [emitted]  
[Main]      images/two-factor/6.png    1.21 kB          [emitted]  
[Main]           images/yubikey.jpg    28.2 kB          [emitted]  
[Main]     locales/bg/messages.json    30.4 kB          [emitted]  
[Main]            images/u2fkey.jpg     178 kB          [emitted]  
[Main]     locales/cs/messages.json    26.7 kB          [emitted]  
[Main]     locales/da/messages.json    25.6 kB          [emitted]  
[Main]     locales/en/messages.json    25.1 kB          [emitted]  
[Main]     locales/de/messages.json    26.7 kB          [emitted]  
[Main]                 package.json  575 bytes          [emitted]  
[Main]     locales/et/messages.json    25.7 kB          [emitted]  
[Main]     locales/fa/messages.json    31.4 kB          [emitted]  
[Main]     locales/fi/messages.json      26 kB          [emitted]  
[Main]     locales/fr/messages.json    27.9 kB          [emitted]  
[Main]     locales/hr/messages.json    26.4 kB          [emitted]  
[Main]     locales/hu/messages.json      27 kB          [emitted]  
[Main]     locales/id/messages.json    25.5 kB          [emitted]  
[Main]     locales/it/messages.json    26.6 kB          [emitted]  
[Main]     locales/ja/messages.json    28.4 kB          [emitted]  
[Main]     locales/ko/messages.json    27.2 kB          [emitted]  
[Main]     locales/nb/messages.json    26.2 kB          [emitted]  
[Main]     locales/nl/messages.json    26.3 kB          [emitted]  
[Main]     locales/pl/messages.json    26.8 kB          [emitted]  
[Main]  locales/pt_BR/messages.json    26.6 kB          [emitted]  
[Main]  locales/pt_PT/messages.json    26.9 kB          [emitted]  
[Main]     locales/ro/messages.json    26.9 kB          [emitted]  
[Main]     locales/ru/messages.json    32.8 kB          [emitted]  
[Main]     locales/sk/messages.json    26.6 kB          [emitted]  
[Main]     locales/sv/messages.json    25.8 kB          [emitted]  
[Main]     locales/th/messages.json    29.8 kB          [emitted]  
[Main]     locales/tr/messages.json    26.6 kB          [emitted]  
[Main]     locales/uk/messages.json    33.2 kB          [emitted]  
[Main]     locales/vi/messages.json    28.5 kB          [emitted]  
[Main]  locales/zh_CN/messages.json    24.5 kB          [emitted]  
[Main]  locales/zh_TW/messages.json    24.5 kB          [emitted]  
[Main]    [4] ./jslib/src/services/constants.service.ts 1.79 kB {0} [built]
[Main]    [7] ./src/main.ts 4.71 kB {0} [built]
[Main]    [8] ./src/services/i18n.service.ts 940 bytes {0} [built]
[Main]   [10] ./src/main/menu.main.ts 20.1 kB {0} [built]
[Main]   [12] ./src/main/messaging.main.ts 2.55 kB {0} [built]
[Main]   [13] ./src/main/powerMonitor.main.ts 2.31 kB {0} [built]
[Main]   [15] ./jslib/src/services/lowdbStorage.service.ts 1.48 kB {0} [built]
[Main]   [18] ./jslib/src/misc/nodeUtils.ts 613 bytes {0} [built]
[Main]   [19] ./jslib/src/misc/utils.ts 9.29 kB {0} [built]
[Main]   [20] ./jslib/src/electron/keytarStorageListener.ts 1.73 kB {0} [built]
[Main]   [22] ./jslib/src/electron/services/electronLog.service.ts 1.44 kB {0} [built]
[Main]   [24] ./jslib/src/electron/services/electronMainMessaging.service.ts 444 bytes {0} [built]
[Main]   [25] ./jslib/src/electron/tray.main.ts 4.46 kB {0} [built]
[Main]   [27] ./jslib/src/electron/updater.main.ts 6.23 kB {0} [built]
[Main]   [29] ./jslib/src/electron/window.main.ts 9.62 kB {0} [built]
[Main]     + 16 hidden modules
[Main] npm info lifecycle bitwarden@0.0.0~postbuild:main: bitwarden@0.0.0
[Main] npm info ok 
[Main] npm run build:main exited with code 0
npm info lifecycle bitwarden@0.0.0~build: Failed to exec build script
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bitwarden@0.0.0 build: `concurrently -n Main,Rend -c yellow,cyan "npm run build:main" "npm run build:renderer"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the bitwarden@0.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /run/build/bitwarden/npm-cache/_logs/2018-08-27T11_45_13_387Z-debug.log
```

The problem seems to originate from the use of `gulp-google-webfonts` to fetch the fonts online, instead of using a local version.

**Comment by ghisvail** [comment]
Finally, thanks to @alexlarsson for his suggestion on IRC:

capture d ecran de 2018-08-28 12-13-10

**Comment by ghisvail** [comment]
bot, build com.bitwarden.desktop

**Comment by ghisvail** [comment]
The current solution bypasses the gulp task responsible for downloading the fonts (`webfonts.patch`) and let `flatpak-builder` unpack the fonts (`webfonts.zip`) and CSS (`webfonts.css`) to the location expected by the rest of the build system.

**Comment by nedrichards** [comment]
bot, build com.bitwarden.desktop

**Comment by ghisvail** [comment]
Bitwarden now builds fine on all arches, finally.

**Comment by nedrichards** [comment]
Great!

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/com.bitwarden.desktop

**Comment by nedrichards** [comment]
@ghisvail is there anyone else from upstream we should add to the repo?

**Comment by ghisvail** [comment]
@nedrichards, please add @kspearrin from upstream.

**Pull request closed without merging**